connectivity/FEATURE_BLE/source/generic/PrivateAddressController.cpp

```diff
@@ -0,0 +1,670 @@
+{
+    if (_generation_started) {
+        return;
+    }
+
+    // non resolvable private address generation has been delayed until now,
+    // generate it.
+    generate_non_resolvable_address();
+
+    _address_rotation_ticker.attach([this] {
+        _event_queue.post([this]{
+            generate_resolvable_address();
+            generate_non_resolvable_address();
+        });
+    }, _rotation_timeout.valueChrono());
+    _generation_started = true;
+}
+
+void PrivateAddressController::stop_private_address_generation()
+{
+    if (_generation_started) {
+        _address_rotation_ticker.detach();
+        _generation_started = false;
+    }
+}
+
+void PrivateAddressController::generate_resolvable_address()
+{
+    if (_local_irk != irk_t{}) {
+        _pal.generate_resolvable_private_address(_local_irk);
+    }
+}
+
+void PrivateAddressController::on_resolvable_private_address_generated(const address_t &rpa)
+{
+    _resolvable_address = rpa;
+    if (_event_handler) {
+        _event_handler->on_resolvable_private_addresses_generated(_resolvable_address);
+    }
+}
+
+void PrivateAddressController::generate_non_resolvable_address()
+{
+    _non_resolvable_address = _pal.generate_non_resolvable_private_address();
+    _event_queue.post([this] {
+        if (_event_handler) {
+            _event_handler->on_non_resolvable_private_addresses_generated(_non_resolvable_address);
+        }
+    });
+}
+
+bool PrivateAddressController::is_controller_privacy_supported()
+{
+    return _pal.is_ll_privacy_supported();
+}
+
+ble_error_t PrivateAddressController::enable_controller_address_resolution(bool enable)
+{
+    MBED_ASSERT(is_controller_privacy_supported());
+    return _pal.set_ll_address_resolution(enable);
+}
+
+uint8_t PrivateAddressController::read_resolving_list_capacity()
+{
+    if (is_controller_privacy_supported()) {
+        return _pal.read_resolving_list_capacity();
+    } else {
+        return RESOLVING_LIST_SIZE;
+    }
+}
+
+ble_error_t PrivateAddressController::add_device_to_resolving_list(
+    target_peer_address_type_t peer_address_type,
+    const address_t &peer_identity_address,
+    const irk_t& peer_irk
+)
+{
+    if (_local_irk == irk_t{}) {
+        return BLE_ERROR_INVALID_STATE;
+    }
+
+    if (is_controller_privacy_supported()) {
+        return queue_add_device_to_resolving_list(
+            peer_address_type,
+            peer_identity_address,
+            peer_irk
+        );
+    } else {
+        for (auto &entry : _resolving_list) {
+            if (entry.populated == false) {
+                entry.peer_address_type = peer_address_type;
+                entry.peer_address = peer_identity_address;
+                entry.peer_irk = peer_irk;
+                entry.populated = true;
+                return BLE_ERROR_NONE;
+            }
+        }
+
+        // Remove unresolved entries from the resolved list
+        remove_resolution_entry_from_cache(
```

Comment:
why now, what has triggered them to become unresolved?

Reply:
Let say as a central you scan and receives an RPA from device X, if you doesn't have an IRK for X then X's RPA is marked as not resolved in the cache.  
The central connects to X, pair and disconnects. After disconnection, X continues to advertise with the same RPA. When asking for resolution, the cache will return that this address is not resolved because the computation has been made before bonding.

To prevent this issue we remove resolution entries that are not associated with an identity when a new device is added to the resolving list. It is just cache invalidation.